### ISSUE TITLE
setup.py --record omits compiled bytecode files when .separate_namespace used

### ISSUE THREAD

**Issue by neirbowj** [Legrandin/pycryptodome#121]
# Test case

```sh
stagedir=`mktemp -d`
recordfile=`mktemp`
touch .separate_namespace
python setup.py install \
    --single-version-externally-managed \
    --root=$stagedir \
    --prefix=/usr/local \
    --record=$recordfile \
    -O1
# compare the output of the following two commands to verify equality
egrep "\.py(c|o)" $recordfile | wc -l
find $stagedir -name "*.pyc" -o -name "*.pyo" | wc -l
rm $recordfile && unset recordfile
rm -fr $stagedir && unset stagedir
```

I ran the test with Python 3.6.4 and Python 2.7.14, both with setuptools 38.2.5. I also verified that this bug does not exist when `.separate_namespace` is absent.

# Expected output

```
324
324
```

# Observed output

```
0
324
```

**Comment by Legrandin** [comment]
Thanks (again) for the very detailed report.
Can you confirm you were using version 3.4.8 (released just today)?
I am trying to reproduce and I obtain:

    324
    324

**Comment by neirbowj** [comment]
```sh
$ git describe --tags
v3.4.8
```

I'll see if I can repro on another system.

**Comment by neirbowj** [comment]
Reproducible on:
* FreeBSD 10.4-STABLE (r325957) under Python 3.6.4 and Python 2.7.14, both with setuptools 38.2.5 (original environment)

Not reproducible on:
* macOS 10.12.6 under Python 3.6.4 and Python 2.7.14, both with setuptools 36.8.0, all from Macports.
* RedHat Enterprise Linux 6.9 under Python 3.6.4 with setuptools 28.8.0 and 38.4.0.

I'll see if I can isolate what's different about the FreeBSD environment.

**Comment by neirbowj** [comment]
The bad interaction appears to be with a FreeBSD-specific patch to all versions of python ([2.7, 3.2-3.5](https://svnweb.freebsd.org/ports?view=revision&revision=405569), [3.6](https://svnweb.freebsd.org/ports/head/lang/python36/files/patch-Lib_distutils_command_install__lib.py?revision=429438&view=markup)) that affects `distutils.command.install_lib:install_lib._bytecode_filenames`.

Perhaps the existence check is either looking at where the bytecode files haven't been moved to yet, or is looking at where the bytecode files were following the move.

**Comment by Legrandin** [comment]
I can also reproduce it now in FreeBSD 11.1, which is not a platform I normally run tests on. I will try to dig up the root case.

**Comment by Legrandin** [comment]
Closing. This should be fixed now in v3.4.9.

**Comment by neirbowj** [comment]
Thank you for working on this. I fear I am doing you no favors by making the test cases too simplistic. Your fix works before but not at or later than python 3.2 because of [PEP 3147](https://www.python.org/dev/peps/pep-3147/).

```sh
$ egrep "\.py(c|o)" $recordfile | sort | head
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/AES.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/AES.pyo
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/ARC2.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/ARC2.pyo
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/ARC4.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/ARC4.pyo
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/Blowfish.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/Blowfish.pyo
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/CAST.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/CAST.pyo
$ find $stagedir -name "*.pyc" -o -name "*.pyo" | sort | sed "s|$stagedir||" | head
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/__pycache__/AES.cpython-36.opt-1.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/__pycache__/AES.cpython-36.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/__pycache__/ARC2.cpython-36.opt-1.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/__pycache__/ARC2.cpython-36.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/__pycache__/ARC4.cpython-36.opt-1.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/__pycache__/ARC4.cpython-36.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/__pycache__/Blowfish.cpython-36.opt-1.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/__pycache__/Blowfish.cpython-36.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/__pycache__/CAST.cpython-36.opt-1.pyc
/usr/local/lib/python3.6/site-packages/Cryptodome/Cipher/__pycache__/CAST.cpython-36.pyc
```

**Comment by Legrandin** [comment]
Thanks again for shooting holes!

Instead of keep second guessing what `distutil` does, I have taken a different approach: I prepare the `Cryptodome` package ahead of time and tell `setup()` to directly work on that.
It's currently on master and a release to pypi will follow shortly - after I test it also on FreeBSD.

**Comment by neirbowj** [comment]
That sounds like a great strategy. Another approach I've been considering is to make `Cryptodome` the native namespace for this package, and to implement what would amount to a dumb redirection wrapper in the `Crypto` namespace. Then `setup.py` could always install `Cryptodome` ('x') and optionally install `Crypto` (not 'x'). Or, for that matter, py-cryptodome could become a separate package implementing the wrapper and depending upon py-cryptodomex as its underlying implementation.

In any case, 12599d6 resolves this bug. Shortly I will update the following proposed patches to the FreeBSD ports tree.

- [security/py-pycryptodome: Prepare for slave port and register CONFLICTS_INSTALL with security/py-pycrypto](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=225216)
- [[NEW PORT] security/py-pycryptodomex: Cryptographic library for Python](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=225217)

**Comment by Legrandin** [comment]
Fixed in v3.4.9.
